doc/designs/quic-design/quic-ackm.md

```diff
@@ -0,0 +1,412 @@
+    /*
+     * Represents an inclusive range of packet numbers [start, end].
+     * start must be <= end.
+     */
+    QUIC_PN start, end;
+} QUIC_ACKM_ACK_RANGE;
+
+typedef struct quic_ackm_ack {
+    /*
+     * A sequence of packet number ranges [[start, end]...].
+     *
+     * The ranges must be sorted in descending order, for example:
+     *      [ 95, 100]
+     *      [ 90,  92]
+     *      etc.
+     *
+     * As such, ack_ranges[0].end is always the highest packet number
+     * being acknowledged and ack_ranges[num_ack_ranges-1].start is
+     * aalways the lowest packet number being acknowledged.
+     *
+     * num_ack_ranges must be greater than zero, as an ACK frame must
+     * acknowledge at least one packet number.
+     */
+    const QUIC_ACKM_ACK_RANGE  *ack_ranges;
+    size_t                      num_ack_ranges;
+
+    OSSL_TIME                   delay_time;
+    uint64_t                    ect0, ect1, ecnce;
+} QUIC_ACKM_ACK;
+
+int QUIC_ACKM_on_rx_ack_frame(QUIC_ACKM *ackm, const QUIC_ACKM_ACK *ack,
+                              int pkt_space, OSSL_TIME rx_time);
+```
+
+### On Packet Space Discarded
+
+This must be called whenever a packet number space is discarded. ACK-tracking
+information for the number space is thrown away. Any previously provided
+`QUIC_ACKM_TX_PKT` structures have their `on_discarded` callback invoked,
+providing an opportunity for them to be freed.
+
+Returns 1 on success.
+
+```c
+int QUIC_ACKM_on_pkt_space_discarded(QUIC_ACKM *ackm, int pkt_space);
+```
+
+### On Handshake Connfirmed
+
+This should be called by the user when the QUIC handshake is confirmed. The PTO
+algorithm behaves differently depending on whether the QUIC handshake is
+confirmed yet.
+
+Returns 1 on success.
+
+```c
+int QUIC_ACKM_on_handshake_confirmed(QUIC_ACKM *ackm);
+```
+
+### On Timeout
+
+This must be called whenever the loss detection deadline expires.
+
+```c
+int QUIC_ACKM_on_timeout(QUIC_ACKM *ackm);
+```
+
+Queries
+-------
+
+These functions allow information about the status of the ACK manager to be
+obtained.
+
+### Get Loss Detection Timeout
+
+This returns a deadline after which `QUIC_ACKM_on_timeout` should be called.
+
+If it is `OSSL_TIME_NEVER`, no timeout is currently active.
+
+The value returned by this function may change after any call to any of the
+event functions above is made.
+
+```c
+OSSL_TIME QUIC_ACKM_get_loss_detection_timeout(QUIC_ACKM *ackm);
+```
+
+### Get ACK Frame
+
+This returns a pointer to a `QUIC_ACKM_ACK` structure representing the
+information which should be packed into an ACK frame and transmitted.
+
+If no new ACK frame is currently needed, returns NULL. After calling this
+function, calling the function immediately again will return NULL.
+
+The structure pointed to by the returned pointer, and the referenced ACK range
+structures, are guaranteed to remain valid until the next call to any
+`QUIC_ACKM` function. After such a call is made, all fields become undefined.
```

Comment:
Synchronisation is assumed to be the responsibility of the caller.

Realistically, the ACKM is a bunch of state and I don't see an easy way to make it thread safe other than  just with a single lock. I believe other implementations are similar in this regard. If there is a way, it would definitely add enough complexity that we should definitely make this work single-threaded first and have it passing tests before we attempt to add that complexity.

Reply:
IMO at worst we could do synchronization via the event queue - i.e., the ack manager would be called only when processing events and things like received ack frames would be created as events from the rx depacketizer.